table/block_based/block_based_table_builder.cc

```diff
@@ -1886,10 +1897,72 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                                         r->pending_handle);
       }
     }
-
+    // TODO: figure out when we can ensure all the buffered data blocks are flushed and call this function
+    // TODO: figure out what is the right status message for successfully flushing all the buffered data blocks
+    if (ok()) {
+      FreeAllBufferedDataBlocskMemWithCache();
+    }
     std::swap(iter, next_block_iter);
   }
   r->data_block_buffers.clear();
+  r->buffer_exceeds_global_memory_limit = false;
+}
+
+// TODO figure out a better name for this function to reflect what it does in
+// the context of this class
+void BlockBasedTableBuilder::ReserveLastBufferedDataBlockMemWithCache() {
+  // TODO-t figure out a better way to represent there is no need to reserve mem due to no cache
+  CacheRep* cache_rep = rep_->cache_rep;
+  if (cache_rep == nullptr) {
```

Comment:
Question 3:
I've seen different practice of condition checking in the codebase. For the scenario here, I can think of three ways of checking against no caching (i.e, when "no_block_cache" is true in the "BlockBasedTableOptions").
(a)check that before calling ReserveLastBufferedDataBlockMemWithCache and use assertion within the function
(b)check inside the function and use early return - like what I am doing here
(c)check inside the function and use "late return" - if (cache_rep != nullptr) {do useful stuff}

I've seen all three practices in the codebase so I figured I should ask if there is a general practice that I should follow in scenario like this.

Reply:
They are all similarly good. I slightly prefer (a). For (b) and (c) the function would more accurately be called `MaybeReserveLastBufferedDataBlockMemInCache()`, with the downside that the reader needs some slight extra effort to learn what cases "maybe" means yes vs. no.